Convert `d` commands to rzshell

b1llow opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

...

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

Partially addresses https://github.com/rizinorg/rizin/issues/1342

b1llow committed:
debug rzshell sketch

XVilka commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -9,16 +9,16 @@ commands:
         summary: Add breakpoint at current offset
         cname: cmd_debug_add_bp
-        args: []
+        args: [ ]
```
Please avoid such unnecessary style changes in YAML files.

b1llow committed:
remove spaces within brackets

b1llow committed:
Merge branch 'dev' into asan-debug-rzshell

b1llow committed:
comment for every handler

b1llow committed:
`dt*`:
  - remove `dtdi`, Because his function is the same as `dtd`

b1llow committed:
dd*

b1llow committed:
fix dd*

b1llow committed:
di* dH dg

b1llow committed:
remove help_msg_dts

b1llow committed:
fix

b1llow committed:
de*
move `"de", " [perm] [rm] [addr|reg|from..to]", "Stop on condition"` to `dee`

b1llow committed:
dL*
move `"dL", " <handler>", "Set debugger handler"` to `dLs`

b1llow committed:
dk*
move `"dk?", "<signal>", "Name/signum resolver"` to `dkr`

b1llow committed:
do*

b1llow committed:
remove code that is no longer used

b1llow committed:
dp*

b1llow committed:
dx*

b1llow committed:
d[wW]*

b1llow committed:
use `RZ_LOG_ERROR` to replace `eprintf`

b1llow committed:
make some `RZ_LOG_ERROR` become rz_cons_printf

b1llow committed:
move `dLs` back to `dL`

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
dcs

b1llow committed:
`dmi*` looks messy needs more refactoring

b1llow committed:
ds*

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -299,8 +299,20 @@ commands:
+              - name: str
+                type: RZ_CMD_ARG_TYPE_STRING
+                optional: true
+          - name: dcs*
```
I would change the name of this command because `*` usually means "rizin output".
`dcst` maybe? @XVilka @wargio ?

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+        args:
+          - name: file
+            type: RZ_CMD_ARG_TYPE_FILE
+            optional: true
```
Split this into `ddl` (list file descriptors) and `dd <file>` (Open and map that file)

I think the "UI" part is wrong? What does it mean?

b1llow replied:
I don't even know what it means.
https://github.com/rizinorg/rizin/blob/b3de5151537aad66aad213caf1443e46c4861394/librz/core/cmd/cmd_debug.c#L71-L72

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+      - name: dd*
+        summary: List file descriptors (in rizin commands)
+        cname: cmd_debug_list_fd
+        args: []
```
when you split `dd` into `ddl` /`dd` you can use `type: ..._STATE` and add OUTPUT_RIZIN as `modes` (search in this file to see how it's done).

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+          - name: fd
+            type: RZ_CMD_ARG_TYPE_NUM
+          - name: offset
+            type: RZ_CMD_ARG_TYPE_NUM
```
```suggestion
            type: RZ_CMD_ARG_TYPE_RZNUM
```
use rz_num_math in the handler.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+          - name: offset
+            type: RZ_CMD_ARG_TYPE_NUM
+          - name: size
+            type: RZ_CMD_ARG_TYPE_NUM
```
```suggestion
          - name: offset
            type: RZ_CMD_ARG_TYPE_RZNUM
          - name: size
            type: RZ_CMD_ARG_TYPE_RZNUM
```
use rz_num_math in the handler.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+          - name: offset
+            type: RZ_CMD_ARG_TYPE_NUM
+          - name: len
+            type: RZ_CMD_ARG_TYPE_NUM
```
```suggestion
          - name: offset
            type: RZ_CMD_ARG_TYPE_RZNUM
          - name: size
            type: RZ_CMD_ARG_TYPE_RZNUM
```
use rz_num_math in the handler.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+        cname: cmd_debug_fd_tty
+        args:
+          - name: ttypath
+            type: RZ_CMD_ARG_TYPE_NUM
```
imo this can be removed as `dd- 0` already does the same and the description of this is a bit misleading.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+  - name: di
+    summary: Show debugger backend information (See dh)
+    subcommands:
+      - name: di
```
please use type: _STATE

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+            optional: true
+          - name: b
+            type: RZ_CMD_ARG_TYPE_FILE
+            optional: true
```
please put this as `diff` in `cmd_shell.yaml` instead of here. the command handler in cmd_shell.c instead of cmd_debug.c

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+            type: RZ_CMD_ARG_TYPE_FILE
+            optional: true
+  - name: dk
+    summary: List, send, get, set, signal handlers of child
```
```suggestion
    summary: Handle signal handlers of child processes
```

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+        summary: >
+          List all signal handlers of child process 
+          / Send KILL signal to child
+          / Set signal handler for <signal> in child
```
split in two commands: `dkl` and `dk <signal>`.

ret2libc replied:
also, you can probably remove the `dk <signal>=<x>` because `rz_debug_kill_setup` seem to not be implemented. For now, just remove that functionality and the empty API please.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+          / Send KILL signal to child
+          / Set signal handler for <signal> in child
+        cname: cmd_debug_signal_list
+        modes:
```
use the _STATE type

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+            type: RZ_CMD_ARG_TYPE_RZNUM
+      - name: dko
+        summary: >
+          List existing signal handling
```
I think `dko` does the same as `dk` so this functionality can be removed.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+              - cont
+  - name: dL
+    summary: List or set debugger handler
+    subcommands:
```
no need for subcommands in this case, just define the command directly

ret2libc replied:
use `type: ..._STATE`

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -312,78 +324,194 @@ commands:
+    summary: List or set debugger handler
+    subcommands:
+      - name: dL
+        summary: List debugger handler
```
```suggestion
        summary: List debugger handler / Set debugger handler <handler>
```

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -459,8 +587,31 @@ commands:
-        cname: cmd_debug_dmi
-        type: RZ_CMD_DESC_TYPE_OLDINPUT
+        subcommands:
+          - name: dmi
```
type: STATE

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -459,8 +587,31 @@ commands:
+            summary: List closest symbol to the current address
+            cname: cmd_debug_dmid
+            args: []
+      #          - name: dmiv
```
why commented?

b1llow replied:
https://github.com/rizinorg/rizin/blob/b3de5151537aad66aad213caf1443e46c4861394/librz/core/cmd/cmd_debug.c#L1232-L1239

Because they were processed together in the original, and then it still use `RZ_MODE_SIMPLE`, `RZ_MODE_JSON`, but now we seem to use `RzOutputMode`.
And I haven't completely changed this part `dmi`.

b1llow replied:
`dmiv` seems to be just an alias for `dmm`

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -527,6 +678,396 @@ commands:
+  - name: do
+    summary: Debug (re)open commands / Open process (reload, alias for 'oo')
+    cname: cmd_debug_process_open
+    type: RZ_CMD_DESC_TYPE_OLDINPUT
```
is this still necessary?

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -527,6 +678,396 @@ commands:
+  - name: dp
+    summary: List, attach to process or thread id
+    subcommands:
+      - name: dp
```
type STATE

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -527,6 +678,396 @@ commands:
+          - name: pid
+            type: RZ_CMD_ARG_TYPE_RZNUM
+            optional: true
+      - name: dpl
```
type STATE

ret2libc reviewed (commented):
Thanks for this, it will really help to have this merged!

I think right now the changes are really a lot and hard to review well though. Would it be feasible for you to split this PR into smaller ones? For example, start by converting `dd` commands, then `dmi`, etc. etc. Otherwise it becomes hard for me and others to review all changes and it may take a very long time to merge this.

It would be awesome to have this in.

b1llow committed:
Update librz/core/cmd_descs/cmd_debug.yaml

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

b1llow committed:
Update librz/core/cmd_descs/cmd_debug.yaml

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

b1llow committed:
Update librz/core/cmd_descs/cmd_debug.yaml

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

b1llow committed:
Update librz/core/cmd_descs/cmd_debug.yaml

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

b1llow committed:
`dcs*` -> `dcst`
move "List file descriptors" to `ddl`
merge `dd*` to `ddl`
`type: RZ_CMD_DESC_TYPE_ARGV_STATE`

b1llow committed:
Merge branch 'asan-debug-rzshell' of https://github.com/rizinorg/rizin into asan-debug-rzshell

b1llow committed:
remove `ddt` `dk <signal>=<x>` `dko` `do`
- use `oo` to replace `do`

b1llow committed:
split `dk` in two commands: `dkl` and `dk <signal>`

b1llow committed:
move `dif` from `cmd_debug` to `cmd_shell` and rename to `diff`

b1llow committed:
move `dmia [<libname> [<symname>]]` to `dmi [-a [<libname> [<symname>]]]`
remove dmiv

b1llow committed:
fix

b1llow committed:
fix

ret2libc commented on librz/core/cmd/cmd_shell.c:
```diff
@@ -165,6 +165,28 @@ RZ_IPI RzCmdStatus rz_cmd_shell_cat_handler(RzCore *core, int argc, const char *
 
+// diff
+RZ_IPI RzCmdStatus rz_cmd_shell_diff_handler(RzCore *core, int argc, const char **argv) {
+	if (argc < 3) {
+		RZ_LOG_ERROR("Cannot open those alias files\n");
+	} else {
```
i don't think it is necessary.

ret2libc commented on librz/core/cmd_descs/cmd_shell.yaml:
```diff
@@ -121,6 +121,16 @@ commands:
+    args:
+      - name: a
+        type: RZ_CMD_ARG_TYPE_FILE
+        optional: true
```
it's not optional

ret2libc commented on librz/core/cmd_descs/cmd_shell.yaml:
```diff
@@ -121,6 +121,16 @@ commands:
+        optional: true
+      - name: b
+        type: RZ_CMD_ARG_TYPE_FILE
+        optional: true
```
it's not optional

ret2libc reviewed (commented):
Any chance you could split this PR in smaller ones?

b1llow commented:
> Any chance you could split this PR in smaller ones?

I want to fix all the tests first.

b1llow committed:
fix

b1llow committed:
fix dmi

XVilka commented:
@imbillow shall we close this one then?

b1llow commented:
> @imbillow shall we close this one then?

yes
